kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp

```diff
@@ -0,0 +1,213 @@
+    if (migrate_to_v4) {
+      store_to(v4_kvbc_->getAdapter());
+    } else {
+      store_to(cat_kvbc_->getAdapter());
+    }
+  });
+  if (std::thread::hardware_concurrency() > 2) {
+    cpu_set_t cpu_set;
+    CPU_ZERO(&cpu_set);
+    CPU_SET(1, &cpu_set);
+    int rc = pthread_setaffinity_np(storage_thread_->native_handle(), sizeof(cpu_set_t), &cpu_set);
+    if (rc != 0) {
+      LOG_WARN(logger_, "Cannot create a CPU affinity for storage thread, so this process might run slow.");
+    }
+  }
+  const auto point_lookup_threads = config_["point-lookup-threads"].as<std::int64_t>();
+  batch_thread_pool_ = std::make_unique<ThreadPool>(static_cast<std::uint32_t>(point_lookup_threads));
+}
+
+void BookKeeper::migrate(const std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& from,
+                         std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& to) {
+  start_batch_id_ = from->getGenesisBlockId();
+  last_batch_id_ = from->getLastBlockId();
+  batch_id_ = 1;
+  std::vector<std::future<void>> tasks;
+  auto point_lookup_batch_size = config_["point-lookup-batch-size"].as<std::int64_t>();
+  for (uint64_t block_id = start_batch_id_; block_id <= last_batch_id_;) {
+    auto last_block_id_for_this_batch =
+        ((block_id + point_lookup_batch_size) <= last_batch_id_) ? block_id + point_lookup_batch_size : last_batch_id_;
+    tasks.push_back(batch_thread_pool_->async(
+        [this](const auto& local_from, auto l_batch_id, auto l_from_block_id, auto l_last_block_id) {
+          read_from(local_from, l_batch_id, l_from_block_id, l_last_block_id);
+        },
+        from,
+        batch_id_.load(),
+        block_id,
+        last_block_id_for_this_batch));
+    ++batch_id_;
+    block_id = last_block_id_for_this_batch + 1;
+  }
+
+  for (auto const& t : tasks) {
+    t.wait();
+  }
+
+  if (storage_thread_->joinable()) {
+    storage_thread_->join();
+  }
+}
+
+void BookKeeper::read_from(const std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& from,
+                           uint64_t batch_id,
+                           uint64_t from_block_id,
+                           uint64_t last_block_id) {
+  auto batch = std::make_shared<std::vector<concord::kvbc::categorization::Updates>>();
+  for (uint64_t block_id = from_block_id; block_id <= last_block_id; ++block_id) {
+    batch->push_back((from->getBlockUpdates(block_id)).value());
+  }
+  {
+    std::unique_lock<std::mutex> lock(batch_mutex_);
+    batches_.emplace(batch_id, batch);
+    batch_cond_var_.notify_all();
+  }
+}
+
+void BookKeeper::store_to(std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& to) {
+  uint64_t curr_batch_id = 1;
+  while (true) {
+    bool wait_pred_result = false;
+    {
+      std::unique_lock<std::mutex> lock(batch_mutex_);
+      auto batch_is_empty = batches_.empty();
+      auto max_batch_id = batch_id_.load();
+      if ((max_batch_id != 0) && batch_is_empty) {
+        return;
```

Comment:
if we return here, if I understand the code, it will stop the storage thread.
Does it mean that the migration is over?

Reply:
Yes this means the migration is over. the logic is updated to take all the cases.